### PR TITLE
Add a distinct error code and description for "main function has wron…

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
…g type"

### CODE REVIEW

File: src/librustc/diagnostics.rs [hunk 1]
```diff
@@ -1661,6 +1661,35 @@ To understand better how closures work in Rust, read:
...
+
+Taking no arguments and returning no type. Example:
+
+```
+fn main() {
+    // your code
+}
+```
+
+If you to get command-line arguments, use `std::env::args`. To exit with a
```
Comment (xen0n) [inline src/librustc/diagnostics.rs]: > If you to [...]

missing verb?

Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Absolutely.

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/librustc/diagnostics.rs [hunk 2]
```diff
@@ -1661,6 +1661,35 @@ To understand better how closures work in Rust, read:
...
+
+Erroneous code example:
+
+```compile_fail,E0579
+fn main() -> i32 { // error: main function has wrong type
+    0
+}
+```
+
+The `main` function prototype should always be the following:
```
Comment (durka) [inline src/librustc/diagnostics.rs]: Rust doesn't have function prototypes so this term is sort of out of the blue. I think showing the correct shape of the function as in the next code example is fine.

Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Less is more I guess. ;)

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by GuillaumeGomez** [comment]
Anyone? 😢

**Comment by jonathandturner** [comment]
@GuillaumeGomez - it looks good, but I was curious - why did the error number change?

```
-fn main() -> i32 { 0 } //~ ERROR E0308
+fn main() -> i32 { 0 } //~ ERROR E0579
```

**Comment by GuillaumeGomez** [comment]
Because the error changed. That's the main purpose of this PR. ;)

**Comment by GuillaumeGomez** [comment]
@jonathandturner: Anything else I need to do?

**Comment by jonathandturner** [comment]
@GuillaumeGomez sorry, I thought you had repurposed an older one.  Looks good.

@bors r+

**Comment by jonathandturner** [comment]
Tidy failures

```
Building stage0 tool tidy (x86_64-unknown-linux-gnu)

   Compiling tidy v0.1.0 (file:///checkout/src/tools/tidy)

    Finished release [optimized] target(s) in 5.27 secs

tidy check (x86_64-unknown-linux-gnu)

duplicate error code: 579

/checkout/src/librustc/diagnostics.rs:1697: E0579: r##"

/checkout/src/librustc_passes/diagnostics.rs:227: E0579: r##"

thread 'main' panicked at 'some tidy checks failed', /checkout/src/tools/tidy/src/main.rs:55

note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by GuillaumeGomez** [comment]
Damn... We waited for too long and a new error code has been added... I'll update the error code number tomorrow...

**Comment by GuillaumeGomez** [comment]
I replaced E0579 with E0580. I'll r+, don't hesitate to r- @jonathandturner if you see anything.

@bors: r+

**Pull request merged**